File: momentra/momentra/Models/GroupModels.swift
import Foundation

enum GroupSplitMode {
    static let equal = "equal"
    static let exact = "exact"
    static let percent = "percent"
    static let shares = "shares"
}

struct GroupMemberSeedIn: Codable {
    let displayName: String
    let email: String?
    let role: String

    enum CodingKeys: String, CodingKey {
        case displayName = "display_name"
        case email
        case role
    }
}

struct GroupMomentCreateIn: Codable {
    let title: String
    let momentType: String
    let targetAmount: Double?
    let destination: String?
    let tripStartDate: String?
    let tripEndDate: String?
    let splitMode: String
    let contributionDueDate: String?
    let members: [GroupMemberSeedIn]
    let rules: GroupMomentRulesIn

    enum CodingKeys: String, CodingKey {
        case title
        case momentType = "moment_type"
        case targetAmount = "target_amount"
        case destination
        case tripStartDate = "trip_start_date"
        case tripEndDate = "trip_end_date"
        case splitMode = "split_mode"
        case contributionDueDate = "contribution_due_date"
        case members
        case rules
    }
}

struct GroupMomentOut: Codable, Identifiable {
    let momentId: String
    /// Organiser Firebase uid; PATCH/DELETE require this to match the signed-in user (server enforces).
    let ownerUid: String?
    let title: String
    let momentType: String?
    let targetAmount: Double?
    let destination: String?
    let contributionDueDate: String?
    let status: String?
    let raisedAmount: Double?
    let splitMode: String?
    let tripStartDate: String?
    let tripEndDate: String?

    var id: String { momentId }

    enum CodingKeys: String, CodingKey {
        case momentId = "moment_id"
        case ownerUid = "owner_uid"
        case title
        case momentType = "moment_type"
        case targetAmount = "target_amount"
        case destination
        case contributionDueDate = "contribution_due_date"
        case status
        case raisedAmount = "raised_amount"
        case splitMode = "split_mode"
        case tripStartDate = "trip_start_date"
        case tripEndDate = "trip_end_date"
    }
}

struct GroupMomentListOut: Codable {
    let moments: [GroupMomentOut]
}

struct GroupMomentRulesOut: Codable {
    let sendPaymentReminders: Bool
    let autoNotifyOnContribution: Bool
    let allowPartialPayments: Bool
    let requireReceiptForExpenses: Bool
    let requireOrganiserApproval: Bool

    enum CodingKeys: String, CodingKey {
        case sendPaymentReminders = "send_payment_reminders"
        case autoNotifyOnContribution = "auto_notify_on_contribution"
        case allowPartialPayments = "allow_partial_payments"
        case requireReceiptForExpenses = "require_receipt_for_expenses"
        case requireOrganiserApproval = "require_organiser_approval"
    }
}

struct GroupMemberLedgerOut: Codable, Identifiable {
    let memberId: String
    let firebaseUid: String?
    let displayName: String?
    let email: String?
    let role: String
    let status: String
    let joinedAt: String?
    let expectedShare: Double
    let contributedTotal: Double
    let paid: Bool

    var id: String { memberId }

    enum CodingKeys: String, CodingKey {
        case memberId = "member_id"
        case firebaseUid = "firebase_uid"
        case displayName = "display_name"
        case email
        case role
        case status
        case joinedAt = "joined_at"
        case expectedShare = "expected_share"
        case contributedTotal = "contributed_total"
        case paid
    }
}

struct GroupBudgetCategoryOut: Codable, Identifiable {
    let categoryId: String
    let categoryKey: String
    let displayName: String
    let capAmount: Double
    let spentAmount: Double

    var id: String { categoryId }

    enum CodingKeys: String, CodingKey {
        case categoryId = "category_id"
        case categoryKey = "category_key"
        case displayName = "display_name"
        case capAmount = "cap_amount"
        case spentAmount = "spent_amount"
    }
}

struct GroupExpenseSplitLineIn: Codable {
    let memberId: String
    let value: Double

    enum CodingKeys: String, CodingKey {
        case memberId = "member_id"
        case value
    }
}

struct GroupExpenseCreateIn: Codable {
    let categoryKey: String
    let subcategory: String?
    let title: String
    let amount: Double
    let expenseDate: String
    let paidByMemberId: String
    let receiptNotes: String?
    let splitMode: String
    let splitLines: [GroupExpenseSplitLineIn]

    enum CodingKeys: String, CodingKey {
        case categoryKey = "category_key"
        case subcategory
        case title
        case amount
        case expenseDate = "expense_date"
        case paidByMemberId = "paid_by_member_id"
        case receiptNotes = "receipt_notes"
        case splitMode = "split_mode"
        case splitLines = "split_lines"
    }
}

struct GroupTotalsOut: Codable {
    let raisedAmount: Double
    let spentExpensesAmount: Double
    let lastActivityAt: String?

    enum CodingKeys: String, CodingKey {
        case raisedAmount = "raised_amount"
        case spentExpensesAmount = "spent_expenses_amount"
        case lastActivityAt = "last_activity_at"
    }
}

struct GroupMomentDetailOut: Decodable {
    let moment: GroupMomentOut
    let members: [GroupMemberLedgerOut]
    let rules: GroupMomentRulesOut
    let budgetCategories: [GroupBudgetCategoryOut]
    let expenses: [GroupExpenseOut]
    let totals: GroupTotalsOut?

    enum CodingKeys: String, CodingKey {
        case moment
        case members
        case rules
        case budgetCategories = "budget_categories"
        case expenses
        case totals
    }
}

struct GroupExpenseOut: Codable, Identifiable {
    let expenseId: String
    let categoryKey: String
    let subcategory: String?
    let title: String
    let amount: Double
    let expenseDate: String
    let paidByMemberId: String?
    let paidByName: String?
    let splitMode: String?
    let status: String
    /// ISO or date string from API; used to pick the row created by the last POST.
    let createdAt: String?

    var id: String { expenseId }

    enum CodingKeys: String, CodingKey {
        case expenseId = "expense_id"
        case categoryKey = "category_key"
        case subcategory
        case title
        case amount
        case expenseDate = "expense_date"
        case paidByMemberId = "paid_by_member_id"
        case paidByName = "paid_by_name"
        case splitMode = "split_mode"
        case status
        case createdAt = "created_at"
    }
}

extension GroupMomentDetailOut {
    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        moment = try c.decode(GroupMomentOut.self, forKey: .moment)
        members = try c.decodeIfPresent([GroupMemberLedgerOut].self, forKey: .members) ?? []
        rules = try c.decode(GroupMomentRulesOut.self, forKey: .rules)
        budgetCategories = try c.decodeIfPresent([GroupBudgetCategoryOut].self, forKey: .budgetCategories) ?? []
        expenses = try c.decodeIfPresent([GroupExpenseOut].self, forKey: .expenses) ?? []
        totals = try c.decodeIfPresent(GroupTotalsOut.self, forKey: .totals)
    }
}

struct GroupInviteLinkOut: Codable {
    let joinUrl: String

    enum CodingKeys: String, CodingKey {
        case joinUrl = "join_url"
    }
}

struct GroupInviteEmailIn: Codable {
    let emails: [String]
    let message: String?
    let resend: Bool
}

struct GroupInviteEmailOut: Codable {
    let sent: Int
    let failed: Int
    let total: Int
    let errorMessages: [String]

    enum CodingKeys: String, CodingKey {
        case sent
        case failed
        case total
        case errorMessages = "error_messages"
    }
}

struct GroupPendingInviteOut: Codable, Identifiable {
    let inviteId: String
    let momentId: String
    let momentTitle: String
    let email: String
    let status: String
    let sentAt: String?
    let createdAt: String?
    let inviteToken: String

    var id: String { inviteId }

    enum CodingKeys: String, CodingKey {
        case inviteId = "invite_id"
        case momentId = "moment_id"
        case momentTitle = "moment_title"
        case email
        case status
        case sentAt = "sent_at"
        case createdAt = "created_at"
        case inviteToken = "invite_token"
    }
}

struct GroupPendingInvitesOut: Codable {
    let invites: [GroupPendingInviteOut]
}

struct GroupMomentRulesIn: Codable {
    let sendPaymentReminders: Bool
    let autoNotifyOnContribution: Bool
    let allowPartialPayments: Bool
    let requireReceiptForExpenses: Bool
    let requireOrganiserApproval: Bool

    enum CodingKeys: String, CodingKey {
        case sendPaymentReminders = "send_payment_reminders"
        case autoNotifyOnContribution = "auto_notify_on_contribution"
        case allowPartialPayments = "allow_partial_payments"
        case requireReceiptForExpenses = "require_receipt_for_expenses"
        case requireOrganiserApproval = "require_organiser_approval"
    }
}

struct GroupMomentPatchIn: Encodable {
    var title: String?
    var targetAmount: Double?
    var destination: String?
    var contributionDueDate: String?
    var rules: GroupMomentRulesIn?
    var status: String?

    enum CodingKeys: String, CodingKey {
        case title
        case targetAmount = "target_amount"
        case destination
        case contributionDueDate = "contribution_due_date"
        case rules
        case status
    }

    func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encodeIfPresent(title, forKey: .title)
        try c.encodeIfPresent(targetAmount, forKey: .targetAmount)
        try c.encodeIfPresent(destination, forKey: .destination)
        try c.encodeIfPresent(contributionDueDate, forKey: .contributionDueDate)
        try c.encodeIfPresent(rules, forKey: .rules)
        try c.encodeIfPresent(status, forKey: .status)
    }
}

// MARK: - Receipt

struct ReceiptUploadOut: Codable {
    let receiptId: String
    let publicUrl: String
    let thumbnailUrl: String?
    let filePath: String
    let mimeType: String?
    let fileSizeBytes: Int?

    enum CodingKeys: String, CodingKey {
        case receiptId = "receipt_id"
        case publicUrl = "public_url"
        case thumbnailUrl = "thumbnail_url"
        case filePath = "file_path"
        case mimeType = "mime_type"
        case fileSizeBytes = "file_size_bytes"
    }
}

File: apk/app/src/main/java/app/momentra/network/group/GroupModels.kt
package app.momentra.network

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/** Allowed values for [GroupMomentCreateIn.splitMode]; must match backend `GROUP_SPLIT_MODE_VALUES`. */
object GroupSplitMode {
    const val EQUAL = "equal"
    const val EXACT = "exact"
    const val PERCENT = "percent"
    const val SHARES = "shares"
}

@Serializable
data class GroupMomentRulesIn(
    @SerialName("send_payment_reminders") val sendPaymentReminders: Boolean = true,
    @SerialName("auto_notify_on_contribution") val autoNotifyOnContribution: Boolean = true,
    @SerialName("allow_partial_payments") val allowPartialPayments: Boolean = true,
    @SerialName("require_receipt_for_expenses") val requireReceiptForExpenses: Boolean = false,
    @SerialName("require_organiser_approval") val requireOrganiserApproval: Boolean = false,
)

@Serializable
data class GroupMemberSeedIn(
    @SerialName("display_name") val displayName: String,
    val email: String? = null,
    val role: String = "member",
)

@Serializable
data class GroupMomentCreateIn(
    val title: String,
    @SerialName("moment_type") val momentType: String = "trip_fund",
    @SerialName("target_amount") val targetAmount: Double? = null,
    val destination: String? = null,
    @SerialName("trip_start_date") val tripStartDate: String? = null,
    @SerialName("trip_end_date") val tripEndDate: String? = null,
    @SerialName("split_mode") val splitMode: String = GroupSplitMode.EQUAL,
    @SerialName("contribution_due_date") val contributionDueDate: String? = null,
    val members: List<GroupMemberSeedIn> = emptyList(),
    val rules: GroupMomentRulesIn = GroupMomentRulesIn(),
)

@Serializable
data class GroupMomentOut(
    @SerialName("moment_id") val momentId: String,
    @SerialName("owner_uid") val ownerUid: String = "",
    val title: String,
    @SerialName("moment_type") val momentType: String,
    @SerialName("target_amount") val targetAmount: Double? = null,
    val destination: String? = null,
    @SerialName("trip_start_date") val tripStartDate: String? = null,
    @SerialName("trip_end_date") val tripEndDate: String? = null,
    @SerialName("split_mode") val splitMode: String = GroupSplitMode.EQUAL,
    @SerialName("contribution_due_date") val contributionDueDate: String? = null,
    val status: String = "active",
    @SerialName("join_url") val joinUrl: String = "",
    @SerialName("joined_count") val joinedCount: Int = 0,
    @SerialName("invited_count") val invitedCount: Int = 0,
    @SerialName("raised_amount") val raisedAmount: Double = 0.0,
)

@Serializable
data class GroupMomentListOut(
    val moments: List<GroupMomentOut> = emptyList(),
)

@Serializable
data class GroupMomentRulesOut(
    @SerialName("send_payment_reminders") val sendPaymentReminders: Boolean = true,
    @SerialName("auto_notify_on_contribution") val autoNotifyOnContribution: Boolean = true,
    @SerialName("allow_partial_payments") val allowPartialPayments: Boolean = true,
    @SerialName("require_receipt_for_expenses") val requireReceiptForExpenses: Boolean = false,
    @SerialName("require_organiser_approval") val requireOrganiserApproval: Boolean = false,
)

@Serializable
data class GroupMemberLedgerOut(
    @SerialName("member_id") val memberId: String,
    @SerialName("firebase_uid") val firebaseUid: String? = null,
    @SerialName("display_name") val displayName: String? = null,
    val email: String? = null,
    val role: String = "",
    val status: String = "",
    @SerialName("joined_at") val joinedAt: String? = null,
    @SerialName("expected_share") val expectedShare: Double = 0.0,
    @SerialName("contributed_total") val contributedTotal: Double = 0.0,
    val paid: Boolean = false,
)

@Serializable
data class GroupBudgetCategoryOut(
    @SerialName("category_id") val categoryId: String,
    @SerialName("category_key") val categoryKey: String,
    @SerialName("display_name") val displayName: String,
    @SerialName("cap_amount") val capAmount: Double = 0.0,
    @SerialName("spent_amount") val spentAmount: Double = 0.0,
)

@Serializable
data class GroupExpenseSplitDetailOut(
    @SerialName("member_id") val memberId: String,
    @SerialName("display_name") val displayName: String? = null,
    val amount: Double,
)

@Serializable
data class GroupExpenseOut(
    @SerialName("expense_id") val expenseId: String,
    @SerialName("category_key") val categoryKey: String,
    val subcategory: String? = null,
    val title: String,
    val amount: Double,
    @SerialName("expense_date") val expenseDate: String,
    @SerialName("paid_by_member_id") val paidByMemberId: String? = null,
    @SerialName("paid_by_name") val paidByName: String? = null,
    @SerialName("split_mode") val splitMode: String = GroupSplitMode.EQUAL,
    @SerialName("split_lines") val splitLines: List<GroupExpenseSplitDetailOut> = emptyList(),
    val status: String = "",
    @SerialName("has_receipt") val hasReceipt: Boolean = false,
    @SerialName("created_at") val createdAt: String? = null,
)

@Serializable
data class GroupTotalsOut(
    @SerialName("raised_amount") val raisedAmount: Double = 0.0,
    @SerialName("spent_expenses_amount") val spentExpensesAmount: Double = 0.0,
    @SerialName("last_activity_at") val lastActivityAt: String? = null,
)

@Serializable
data class GroupExpenseSplitLineIn(
    @SerialName("member_id") val memberId: String,
    val value: Double,
)

@Serializable
data class GroupExpenseCreateIn(
    @SerialName("category_key") val categoryKey: String,
    val subcategory: String? = null,
    val title: String,
    val amount: Double,
    @SerialName("expense_date") val expenseDate: String,
    @SerialName("paid_by_member_id") val paidByMemberId: String,
    @SerialName("receipt_notes") val receiptNotes: String? = null,
    @SerialName("split_mode") val splitMode: String = GroupSplitMode.EQUAL,
    @SerialName("split_lines") val splitLines: List<GroupExpenseSplitLineIn> = emptyList(),
)

@Serializable
data class GroupInviteLinkOut(
    @SerialName("join_url") val joinUrl: String,
)

@Serializable
data class GroupInviteEmailIn(
    val emails: List<String> = emptyList(),
    val message: String? = null,
    val resend: Boolean = false,
)

@Serializable
data class GroupInviteEmailOut(
    val sent: Int = 0,
    val failed: Int = 0,
    val total: Int = 0,
    @SerialName("error_messages") val errorMessages: List<String> = emptyList(),
)

@Serializable
data class GroupPendingInviteOut(
    @SerialName("invite_id") val inviteId: String,
    @SerialName("moment_id") val momentId: String,
    @SerialName("moment_title") val momentTitle: String,
    val email: String,
    val status: String,
    @SerialName("sent_at") val sentAt: String? = null,
    @SerialName("created_at") val createdAt: String? = null,
    @SerialName("invite_token") val inviteToken: String,
)

@Serializable
data class GroupPendingInvitesOut(
    val invites: List<GroupPendingInviteOut> = emptyList(),
)

@Serializable
data class GroupMomentDetailOut(
    val moment: GroupMomentOut,
    val members: List<GroupMemberLedgerOut> = emptyList(),
    val rules: GroupMomentRulesOut = GroupMomentRulesOut(),
    @SerialName("budget_categories") val budgetCategories: List<GroupBudgetCategoryOut> = emptyList(),
    val expenses: List<GroupExpenseOut> = emptyList(),
    val totals: GroupTotalsOut? = null,
)

@Serializable
data class GroupMomentPatchIn(
    val title: String? = null,
    @SerialName("target_amount") val targetAmount: Double? = null,
    val destination: String? = null,
    @SerialName("contribution_due_date") val contributionDueDate: String? = null,
    val rules: GroupMomentRulesIn? = null,
    val status: String? = null,
)

